### PR TITLE
Polling message

### PR DESCRIPTION
**Pull request by afiebig**
- Add a function to execute a Polling request to the POS.
- Add several utilities using libserialport to connect and communicate with the POS.

### CODE REVIEW

File: src/transbank.c [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+  struct sp_port **ports;
+  char* portList;
+
+  int retval = sp_list_ports(&ports);
+
+  if (retval == SP_OK) {
+    int separators = 0;
+    int chars = 0;
+    for (int i=0; ports[i] != NULL; i++){
+      separators++;
```
Comment (a1b4) [inline src/transbank.c]: Can not use `separators` as the `i` in the for?
If the answer is no, maybe you can do `for (int i=0; ports[i] != NULL; i++, separators++)`

File: src/transbank.c [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+int load_keys(){
+  int tries = 0;
+  do
+  {
+    int retval = write_message(LOAD_KEYS);
+    if (retval == TBK_OK){
+      char* returnBuf;
+      retval = read_ack();
+      if (retval == TBK_OK && returnBuf[0] == ACK){
```
Comment (a1b4) [inline src/transbank.c]: returnBuf will never be empty?

File: README.md [hunk 3]
```diff
@@ -1,37 +1,36 @@
...
+- Visual Studio 2017 + Complemento para Desarrollo escritorio c++
+- Driver USB Serial Chipset CH340: <http://www.wch.cn/download/CH341SER_EXE.html>
+- [msys2 - mingw-w64](http://www.msys2.org/) Seguir las instrucciones en el sitio.
+    - instalar tolchain de mingw
+        - 32bits: pacman -S mingw-w64-i686-toolchain
+        - 64bits: pacman -S mingw-w64-x86_64-toolchain
+- Swig (puedes usar Chocolatey para instalarlo)
 
-## Dev
+### UniX
```
Comment (RodrigoAyala) [inline README.md]: Typo, It should be Unix or UNIX

File: README.md [hunk 4]
```diff
@@ -1,37 +1,36 @@
...
-
-`external/`: Directory for packages/projects to be used by the project, but not edited as part of the project.
+## Dev
 
-`extras/`: Directory containing extra/optional submodules for the project.
+- git clone git://sigrok.org/libserialport (Also a Dependencie)
+- Driver USB Serial Chipset Prolific: <http://www.prolific.com.tw/US/ShowProduct.aspx?p_id=229&pcid=41>
 
-`data/`: Directory containing non-source code aspects of the project. This might include graphics and markup files.
+### osx
```
Comment (RodrigoAyala) [inline README.md]: The proper name is MacOS

File: README.md [hunk 5]
```diff
@@ -1,37 +1,36 @@
...
 
-`tools/`: Directory containing development utilities, such as build and refactoring scripts
+- brew install automake
```
Comment (RodrigoAyala) [inline README.md]: This requires to have HomeBrew installed, maybe you should add a "requirements" step mentioning this.

File: README.md [hunk 6]
```diff
@@ -42,3 +41,7 @@ cc src/hello.c -o build/hello -lserialport
 ### RUN
 
 ./build/hello
+
+### Install
```
Comment (RodrigoAyala) [inline README.md]: There is a mix between English and Spanish, please stay with Spanish :)

File: examples/main.c [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        //if (get_totals() == TBK_OK)
+        //  printf("Totals OK\n");
+        //else
+        //  printf("Unable to get Totals\n");
+
+        //printf("Change to Normal Pos. \n\n");
+        //if (change_normal_mode() == TBK_OK)
+        //  printf("Success\n");
+        //else
+        //  printf("Nothing to do\n");
```
[the remark this answers is not in the document]
Comment (RodrigoAyala) [reply examples/main.c]: +1 to @a1b4's comment

File: src/transbank.c [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+struct message_t {
+  int payloadSize;
+  int responseSize;
+  int retries;
+  char* payload;
+};
+
+static Message GET_TOTALS = {
+  .payload = GET_TOTALS_MESSAGE,
+  .payloadSize =7,
```
Comment (RodrigoAyala) [inline src/transbank.c]: `=7` should be `= 7`

File: src/transbank.c [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+    if (portList != NULL ){
+      for (int i = 0; i < separators; i++) {
+        strcat(portList, sp_get_port_name(ports[i]));
+        if (i < separators -1)
+        {
+          strcat(portList, "|");
+        }
+      }
+    }
+  }else {
```
Comment (RodrigoAyala) [inline src/transbank.c]: typo `}else {` should be `} else {`

File: src/transbank.c [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+char* list_ports() {
+  struct sp_port **ports;
+  char* portList;
+
+  int retval = sp_list_ports(&ports);
+
+  if (retval == SP_OK) {
+    int separators = 0;
+    int chars = 0;
+    for (int i=0; ports[i] != NULL; i++){
```
Comment (RodrigoAyala) [inline src/transbank.c]: ⚠️ This is a Question and maybe a suggestion, but you can ignore it:

I saw that you're using libserialport. Therefore, sp_list_ports fills `ports` with an array of pointers to sp_port structures, terminated by a NULL. 

In that case, `sizeof(ports)/sizeof(ports[0]);` could return a possible value for separators? (I don't know if the NULL element is malloc'ed, affecting the `sizeof(ports)` result.

File: src/transbank.c [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+  }else {
+    char* error = "No serial devices detected\n";
+    portList = malloc(strlen(error) * sizeof(char*));
+    strcpy(portList, error);
+  }
+  sp_free_port_list(ports);
+  return portList;
+}
+
+char* get_configured_port_name(){
```
Comment (RodrigoAyala) [inline src/transbank.c]: Watch out, there are some function declarations with `(){` and other with a space like `() {`.

File: src/transbank.c (hunk 1 above)
Comment (afiebig) [reply src/transbank.c]: I can surely use separators as the counter, i just split it in order to make clearer that i was counting the loops to determine the number of separator to allocate the memory.

File: src/transbank.c [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+  do
+  {
+    int retval = write_message(LOAD_KEYS);
+    if (retval == TBK_OK){
+      char* returnBuf;
+      retval = read_ack();
+      if (retval == TBK_OK && returnBuf[0] == ACK){
+        retval = read_bytes(returnBuf, LOAD_KEYS);
+        if (retval == TBK_OK)
+          for(int i=0; i < sizeof(returnBuf); i++)
```
Comment (RodrigoAyala) [inline src/transbank.c]: This line code is fine, but beware that avoiding using curly braces could lead to a misunderstood when modifying the code in the future.

As a curiosity, Apple's SSL bug was avoidable using an always-braces rule.

Take it or leave it :)

File: src/transbank.c (hunk 10 above)
Comment (afiebig) [reply src/transbank.c]: It works if the structure size is fixed, but serial-port names are string with different sizes.

**Pull request merged**